File: services/policyd/testing/stub/node/src/create-identity-metadata.ts
import type {
  Metadata
} from "@grpc/grpc-js";
import {
  createIdentityWorkloadMetadata
} from "./create-identity-workload-metadata.js";
import type {
  IdentityCallContext
} from "./identity-call-context.js";

export async function createIdentityMetadata(
  workloadTokenPath: string,
  context: IdentityCallContext
): Promise<Metadata> {
  const metadata = await createIdentityWorkloadMetadata(
    workloadTokenPath,
    context.traceparent);
  metadata.add(
    "ctlflow-invocation",
    `Bearer ${context.invocationToken}`);
  return metadata;
}

File: services/policyd/csharp/src/CtlFlow.Policy.Policyd.Service/Identity/CreateIdentityMetadata.cs
using System.Diagnostics;
using CtlFlow.Policy.Policyd.Service.Configuration;
using CtlFlow.Policy.Policyd.Service.Security.Invocations;
using CtlFlow.Policy.Policyd.Service.Telemetry;
using Grpc.Core;

namespace CtlFlow.Policy.Policyd.Service.Identity;

internal static partial class IdentityFacts
{
    internal static async Task<Metadata> CreateIdentityMetadata(
        IdentitySettings settings,
        InvocationToken invocation,
        Activity? activity,
        CancellationToken cancellation)
    {
        var token = (await File.ReadAllTextAsync(
            settings.WorkloadTokenFilePath,
            cancellation)).Trim();
        if (token.Length is < 1 or > 16_384)
        {
            throw new IdentityUnavailableException(
                new InvalidDataException(
                    "The identity workload token is invalid"));
        }
        var headers = new Metadata
        {
            { "authorization", $"Bearer {token}" },
            {
                "ctlflow-invocation",
                $"Bearer {invocation.ReadForIdentityForwarding()}"
            }
        };
        PolicydTelemetry.AddTraceContext(headers, activity);
        return headers;
    }
}
